File: 在线仓库/modelzoo-master/samples/built-in/classification/Vit-B-16/script/vit_preprocess.py
# Copyright (c) ModelZoo. 2025-2025. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os, argparse, cv2
import numpy as np
from tqdm import tqdm
from PIL import Image
import multiprocessing

model_config = {
    384: {'resize': 384, 'centercrop': 384, 'mean': [0.5, 0.5, 0.5], 'std': [0.5, 0.5, 0.5]},  # resize 384
    224: {'resize': 248, 'centercrop': 224, 'mean': [0.5, 0.5, 0.5], 'std': [0.5, 0.5, 0.5]},  # resize 224
}

# 获取预处理后的bin图片
def get_preprocessed_bin(img_batches, batch, store_path, image_size):
    # 图像缩放
    def img_resize(img, size, interpolation=Image.BICUBIC):
        if not isinstance(size, int):
            return img.resize(size[::-1], interpolation)

        width, height = img.size
        if (width <= height and width == size) or (height <= width and height == size):
            return img

        if width < height:
            reWidth = size
            reHeight = int(size * height / width)
            return img.resize((reWidth, reHeight), interpolation)
        reHeight = size
        reWidth = int(size * width / height)
        ret_img =  img.resize((reWidth, reHeight), interpolation)
        return ret_img
    
    # 中心裁剪    
    def crop_center(img, output_size):
        output_size = (int(output_size), int(output_size))
        img_width, img_height = img.size
        crop_height, crop_width = output_size
        top = int(round((img_height - crop_height) / 2.))
        left = int(round((img_width - crop_width) / 2.))
        ret_img = img.crop((left, top, left + crop_width, top + crop_height))
        return ret_img

    # 归一化
    def normalization(pic, avgs, varis):
        pic = np.array(pic)
        for chn, (avg, vari) in enumerate(zip(avgs, varis)):
            pic[:, :, chn] = (pic[:, :, chn] - avg) / vari
        return pic
    # 遍历图像批次

    for filename, filename_path in tqdm(img_batches[batch]):
        pic = Image.open(filename_path).convert('RGB')
        # 这里可以将image 转成f32 类型的吗，能否减少下面resize 的精度损失
        pic = img_resize(pic, model_config[image_size]['resize']) # Resize
        pic = crop_center(pic, model_config[image_size]['centercrop']) # CenterCrop

        # 转换成浮点数，并归一到[0,1]之间
        pic = np.array(pic).astype(np.float32) / 255

        # 均值归一化
        pic = normalization(pic, model_config[image_size]['mean'], model_config[image_size]['std']) # Normalization
        # 将HWC 调整成 CHW 排布
        pic = pic.transpose((2, 0, 1))
        # 保存预处理后的结果 bin
        pic.tofile(os.path.join(store_path, filename.split('.')[0] + ".bin"))

def preprocess_data(data_path, store_path, image_size):
    # 文件列表
    print("preprocess_data doing...")
    file_list = os.listdir(data_path)
    img_list_info = []
    for filename in file_list:
        filename_path = os.path.join(data_path, filename)
        if os.path.isdir(filename_path):
            img_list_info += \
                [(image_name, os.path.join(filename_path, image_name)) for image_name in os.listdir(filename_path)]
        else:
            img_list_info.append((filename, filename_path))
    # 单个线程处理的批次大小，相应的有10个线程处理
    thread_batchsize = 5000
    img_list_info.sort()
    if len(img_list_info) < thread_batchsize:
        img_batches = [img_list_info[0 : len(img_list_info)]]
    else:
        #单个批次的图片数组
        img_batches = [img_list_info[i:i + thread_batchsize] for i in range(0, len(img_list_info), thread_batchsize) if img_list_info[i:i + thread_batchsize] != []]

    thrd_pool = multiprocessing.Pool(len(img_batches))
    for img_batch in range(len(img_batches)):
        thrd_pool.apply_async(get_preprocessed_bin, args=(img_batches, img_batch, store_path, image_size))
    # 关闭线程池
    thrd_pool.close()
    thrd_pool.join()
    print("process done.")


if __name__ == "__main__":
    # 解析参数
    parser_args = argparse.ArgumentParser(description='Path', add_help=False)
    # 预处理后的路径
    parser_args.add_argument('--store_path', default=r"data/img", metavar='DIR',  help='Save the preprocessed results')
    # 源数据路径
    parser_args.add_argument('--data_path', default=r"../../../../datasets/ImageNet/val", metavar='DIR',
        help='Original dataset')
    # 预处理后的大小
    parser_args.add_argument('--image_size', type=int, required=True, help='Size of the preprocessed image.')

    args = parser_args.parse_args()
    # 创建文件夹
    if not os.path.exists(args.store_path):
        os.makedirs(args.store_path)
    # 预处理
    preprocess_data(args.data_path, args.store_path, args.image_size)

File: 在线仓库/modelzoo-master/samples/common/infer/preprocess/vit_preprocess.cpp
/*
 * Copyright (c) ModelZoo. 2025-2025. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


#include "vit_preprocess.h"
#include "log.h"
#include "dev_interface_adapter.h"
#include <opencv2/opencv.hpp>
#include "PillowResize/PillowResize.hpp"
#include <string>
#include <unordered_map>
#include <algorithm>
#include <regex>
#include <fstream>
#include "utils.h"

namespace Infer {
static cv::Mat PillowResizeKeepRatio(const cv::Mat& img, int targetSize, int32_t interpolation = PillowResize::INTERPOLATION_BILINEAR)
{
    if (targetSize <= 0) {
        return img;
    }
    int width = img.cols;   // 图像宽度
    int height = img.rows;  // 图像高度
    // 检查是否已满足目标尺寸条件
    if ((width <= height && width == targetSize) || (height <= width && height == targetSize)) {
        return img;
    }
    
    cv::Mat targetImg;
    if (width < height) {
        int newWidth = targetSize;
        int newHeight = static_cast<int>(height * static_cast<float>(targetSize) / width);
        targetImg = PillowResize::resize(img, cv::Size(newWidth, newHeight), interpolation);
    } else {
        int newHeight = targetSize;
        int newWidth = static_cast<int>(width * static_cast<float>(targetSize) / height);
        targetImg = PillowResize::resize(img, cv::Size(newWidth, newHeight), interpolation);
    }

    return targetImg;
}

static cv::Mat CenterCrop(const cv::Mat& img, int cropSize)
{
    if (cropSize <= 0) {
        return img;
    }
    int width = img.cols;
    int height = img.rows;
    
    // 计算裁剪区域
    int startX = std::max(0, (width - cropSize) / 2);
    int startY = std::max(0, (height - cropSize) / 2);
    int cropWidth = std::min(cropSize, width - startX);
    int cropHeight = std::min(cropSize, height - startY);
    
    return img(cv::Rect(startX, startY, cropWidth, cropHeight)).clone();
}

static cv::Mat HwcToChw(const cv::Mat& hwcImg)
{
    CV_Assert(!hwcImg.empty() && hwcImg.channels() == 3);
    std::vector<cv::Mat> channels;
    cv::split(hwcImg, channels);

    cv::Mat chwMat(3, hwcImg.rows * hwcImg.cols, hwcImg.depth());
    chwMat = cv::Scalar(0);

    for (int channelIdx = 0; channelIdx < 3; ++channelIdx) {
        cv::Mat flatChannel = channels[channelIdx].reshape(1, 1).clone();
        flatChannel.copyTo(chwMat.row(channelIdx));
    }

    return chwMat; // 返回 3 × (高度*宽度) 的CHW矩阵
}

static int SelectDtype(const std::string& typeStr)
{
    static const std::unordered_map<std::string, int> typeMap = {
        {"uint8",   CV_8U},   // 无符号8位整数
        {"int8",    CV_8S},   // 有符号8位整数
        {"int16",   CV_16S},  // 有符号16位整数
        {"int32",   CV_32S},  // 有符号32位整数
        {"float32", CV_32F}   // 单精度浮点数
    };
    auto it = typeMap.find(typeStr);
    return (it != typeMap.end()) ? it->second : CV_8U; 
}

static std::string GetFileExtension(const std::string& filename)
{
    size_t dotPos = filename.find_last_of('.');
    if (dotPos == std::string::npos) {
        return "";
    }
    std::string ext = filename.substr(dotPos + 1);
    std::transform(ext.begin(), ext.end(), ext.begin(), ::tolower);
    return ext;
}

static bool ParseYuvDimensions(const std::string& filename, int& width, int& height)
{
    // 正则表达式匹配 "_数字*数字" 模式
    std::regex pattern(R"(_(\d+)\*(\d+))");
    std::smatch matches;
    
    if (std::regex_search(filename, matches, pattern) && matches.size() == 3) {
        try {
            width = std::stoi(matches[1].str());
            height = std::stoi(matches[2].str());
            return (width > 0 && height > 0);
        } catch (...) {
            return false;
        }
    }
    return false;
}

static bool Yuv420SpToRgb(const unsigned char* yuvData, int width, int height, cv::Mat& rgbMat)
{
    // 输入参数有效性检查
    if (yuvData == nullptr || width <= 0 || height <= 0) {
        return false;
    }
    
    int ySize = width * height;
    int uvSize = ySize / 2;

    // 创建YUV I420格式的Mat
    cv::Mat yuvI420(height + height/2, width, CV_8UC1);
    if (yuvI420.empty()) {
        return false;
    }
    // 复制Y通道
    memcpy(yuvI420.data, yuvData, ySize);
    // 分离VU交错数据为U和V平面
    unsigned char* uPtr = yuvI420.data + ySize;
    unsigned char* vPtr = uPtr + uvSize/2;
    const unsigned char* vuPtr = yuvData + ySize;
    
    for (int i = 0; i < uvSize / 2; ++i) {
        vPtr[i] = vuPtr[2 * i];     // V分量
        uPtr[i] = vuPtr[2 * i + 1]; // U分量
    }

    cv::cvtColor(yuvI420, rgbMat, cv::COLOR_YUV2RGB_I420);
    // 检查输出矩阵是否有效
    return !rgbMat.empty();
}

static bool ConvertImageToRgb(const std::string& filename, cv::Mat& rgbMat)
{
    std::string ext = GetFileExtension(filename);
    // 处理常见图像格式: JPEG/JPG/PNG/BMP
    if (ext == "jpg" || ext == "jpeg" || ext == "png" || ext == "bmp") {
        // 使用OpenCV直接读取
        cv::Mat img = cv::imread(filename, cv::IMREAD_UNCHANGED);
        if (img.empty()) {
            LOG(ERROR) << "fail to imread, path : " << filename;
            return false;
        }
        
        // 转换为RGB格式 (OpenCV默认读取为BGR)
        if (img.channels() == 3) {
            cv::cvtColor(img, rgbMat, cv::COLOR_BGR2RGB);
        } else if (img.channels() == 4) {
            // 处理带Alpha通道的图像
            cv::cvtColor(img, rgbMat, cv::COLOR_BGRA2RGB);
        } else if (img.channels() == 1) {
            // 处理灰度图像
            cv::cvtColor(img, rgbMat, cv::COLOR_GRAY2RGB);
        } else {
            return false; // 不支持的通道数
        }
    } else if (ext == "yuv420sp") {
        int yuvWidth = 0;
        int yuvHeight = 0;
        // 从文件名解析宽高
        if (!ParseYuvDimensions(filename, yuvWidth, yuvHeight)) {
            return false; // 无法解析宽高
        }
        
        // 计算YUV数据大小
        int ySize = yuvWidth * yuvHeight;
        int uvSize = ySize / 2;
        int totalSize = ySize + uvSize;
        
        // 读取YUV数据
        std::vector<unsigned char> yuvData(totalSize);
        std::ifstream file(filename, std::ios::binary);
        if (!file.read(reinterpret_cast<char*>(yuvData.data()), totalSize)) {
            return false;
        }
        
        // 转换为RGB
        Yuv420SpToRgb(yuvData.data(), yuvWidth, yuvHeight, rgbMat);
    }
    else {
        LOG(ERROR) << "fail to convert image to rgb";
        return false; // 不支持的格式
    }
    return true;
}

/**
 * @brief 图像归一化函数（对应Python版本逻辑）
 * @param pic 输入图像（OpenCV Mat，格式为 HWC，通道顺序BGR/RGB均可，数据类型建议float32）
 * @param avgs 各通道的均值数组（顺序需和图像通道对应）
 * @param varis 各通道的方差数组（顺序需和图像通道对应）
 * @return 归一化后的图像（Mat，格式和输入一致）
 */
static void Normalization(cv::Mat& pic, const std::vector<float>& avgs, const std::vector<float>& varis) {
    // 1. 输入合法性校验
    if (pic.empty()) {
        throw std::invalid_argument("Error: Input image is empty!");
    }
    if (avgs.size() != varis.size()) {
        throw std::invalid_argument("Error: avgs and varis size must be equal!");
    }
    const int channelNum = pic.channels();
    if ((int)avgs.size() != channelNum) {
        throw std::invalid_argument(
            "Error: avgs/varis size(" + std::to_string(avgs.size()) + 
            ") not match image channels(" + std::to_string(channelNum) + ")!"
        );
    }
    // 2. 第二步：按通道执行 (val - avg) / vari 归一化（原地修改）
    // 优化：指针遍历，高效操作
    for (int chn = 0; chn < channelNum; ++chn) {
        const float avg = avgs[chn];
        const float vari = varis[chn];
        
        // 遍历每一行
        for (int row = 0; row < pic.rows; ++row) {
            float* rowPtr = pic.ptr<float>(row); // 获取当前行指针
            // 遍历每一列的当前通道
            for (int col = 0; col < pic.cols; ++col) {
                const int idx = col * channelNum + chn; // 计算像素索引
                // 原地修改：(缩放后的值 - 均值) / 方差
                rowPtr[idx] = (rowPtr[idx] - avg) / vari;
            }
        }
    }
}

static cv::Mat ProcessSingleImage(const std::string& imgPath, const VitPrerocessOptions& options) 
{
    cv::Mat rgbMat;
    if (!ConvertImageToRgb(imgPath, rgbMat)) { 
        LOG(ERROR) << "fail to convert image to rgb";
        return {};
    }
    cv::Mat image = CenterCrop(PillowResizeKeepRatio(rgbMat, options.resizeValue, options.interpolationValue), options.centerCropValue);
    // image.convertTo(image, SelectDtype(options.dtypeStr));
    image.convertTo(image, SelectDtype("float32"), 1.0 / 255.0); // 临时修改适配 vit dpico, 转f32, 并做归一化
    std::vector<float> avgs = {0.5f, 0.5f, 0.5f};  
    std::vector<float> varis = {0.5f, 0.5f, 0.5f};
    Normalization(image, avgs, varis);

    if (options.chwFlag) {
        image = HwcToChw(image);
    }
    return image.clone(); // 确保内存连续
}

bool VitPreprocess(std::vector<std::string>& fileList, std::vector<TensorBuf>& tensorBufs, std::vector<TensorDesc>& tensorDescs, const VitPrerocessOptions& options)
{
    for (size_t i = 0; i < fileList.size(); ++i) {
        cv::Mat image = ProcessSingleImage(fileList[i], options);
        if (image.empty()) {
            LOG(ERROR) << "fail to preprocess " << fileList[i];
            return false;
        }
        size_t imageSize = image.total() * image.elemSize();
        if (!image.isContinuous()) {
            image = image.clone();
        }
        if (!PadDataToTensorBuf(image.data, imageSize, tensorDescs[i], tensorBufs[i])) {
            LOG(ERROR) << "fail to pad data";
            return false;
        }
    }
    return true;
}
}